wagtail/contrib/search_promotions/views.py

```diff
@@ -2,85 +2,104 @@
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                    placeholder=_("Search promoted results"),
-                ),
-            },
-        )
+    ordering = "query_string"
+
+    def get_ordering(self):
+        # Ordering
+        valid_ordering = ["query_string", "-query_string", "views", "-views"]
+        ordering = self.ordering
+
+        if (
+            "ordering" in self.request.GET
+            and self.request.GET["ordering"] in valid_ordering
+        ):
+            ordering = self.request.GET["ordering"]
+        return ordering
+
+    def get(self, request):
+        # Query
+        queries = Query.objects.filter(editors_picks__isnull=False).distinct()
+        ordering = self.get_ordering()
+        if "views" in ordering:
+            queries = queries.annotate(
+                views=functions.Coalesce(Sum("daily_hits__hits"), 0)
+            )
+        queries = queries.order_by(ordering)
+
+        # Search
+        query_string = self.request.GET.get("q", "")
+        if query_string:
+            queries = queries.filter(query_string__icontains=query_string)
+            self.is_searching = True
+
+        # Paginate
+        paginator = Paginator(queries, per_page=self.paginate_by)
+        page_number = self.request.GET.get("p", 1)
+        try:
+            queries = paginator.page(page_number)
+        except InvalidPage:
+            raise Http404
+
+        if request.headers.get("x-requested-with") == "XMLHttpRequest":
+            return render(
+                request,
+                self.result_template_name,
+                {
+                    "is_searching": self.is_searching,
+                    "ordering": self.get_ordering(),
+                    "queries": queries,
+                    "query_string": query_string,
+                },
+            )
+        else:
+            return render(
+                request,
+                self.template_name,
+                {
+                    "is_searching": self.is_searching,
+                    "ordering": self.get_ordering(),
+                    "queries": queries,
+                    "query_string": query_string,
+                    "search_form": SearchForm(
+                        data={"q": query_string} if query_string else None,
+                        placeholder=_("Search promoted results"),
+                    ),
+                },
```

Comment:
We can probably just use the built in `get_context_data` for 90% of what we need here, it already will include `is_searching` and `query_string` and most of these things. even the search form.

We will need to have logic to avoid passing some context when x-requested-with is used though.